core/src/main/java/smile/regression/treeshap/ShapCalculationTemp.java

```diff
@@ -0,0 +1,62 @@
+package smile.regression.treeshap;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import smile.util.MutableInt;
+
+/** 
+ * recyclable class as a register during shap calculation to avoid mass instantiation in memory 
+
+ * <pre>
+ * https://github.com/slundberg/shap/blob/master/shap/explainers/pytree.py
+ * </pre>
+ * 
+ * @author ray
+ */
+public class ShapCalculationTemp {
+
+  public int s;
+  public List<MutableInt> feature_indexes;
```

Comment:
These lists are already get pre-allocated with proper size estimate and never need re-creation on the fly. 

But since Java do not have a convenient and performance-friendly sub-range support for plain array type, while List class do have such support via `subList` view which is preferred here to avoid complex implementation code (otherwise we might need manually handle the subrange indexing into the plain array and modify many functions accordingly in the algorithm to accept subrange indexes as we traverse down the tree)